### PR TITLE
Initial fork to support one-time passwords (OTPs)

### PR DESCRIPTION
**Pull request by anandaroop**
https://artsyproduct.atlassian.net/browse/TRUST-209

tl;dr this PR represents:

- a fork of the [`passport-local` authentication strategy](https://github.com/jaredhanson/passport-local) that Force currently [relies on](https://github.com/artsy/artsy-passport/blob/4c18f2446e56eca1562842ea7467ebf71d0de055/lib/passport/index.coffee#L16-L22) to authenticate with Gravity
- an adaptation of that strategy to allow for an optional OTP field to be passed through in the authentication request to Gravity

### Background

 - Force, along with some other node client apps, uses https://github.com/artsy/artsy-passport to set up shared authentication logic and endpoints
- artsy-passport in turn sits on top of [Passport](http://www.passportjs.org) ([repo](https://github.com/jaredhanson/passport)), an authentication framework that allows for hundreds of  [pluggable authentication strategies](http://www.passportjs.org/packages/)
- artsy-passport configures the [passport-local](https://github.com/jaredhanson/passport-local) strategy (as well as passport-facebook and passport-apple for 3rd-party oauth)

### Problem 

The passport-local strategy assumes authentication via a [simple username/password pair](https://github.com/jaredhanson/passport-local/blob/2bf3939ca369e08a47a28585c2ccfb3cecffeb9c/lib/strategy.js#L63-L95) of credentials, with no support for second factors, such as the `otp_attempt` parameter that Gravity now expects when enforcing 2fa.

Thus we'd need to have a slightly different strategy to support our use case of enforcing 2fa within Force.

### Alternative solutions (i.e. why fork?)

Other otp-aware strategies do exist within the Passport ecosystem, but they do not appear to be a good fit for our system design.

|lib|notes|
|---|---|
|[passport-otp](https://github.com/jaredhanson/passport-otp)|Only accepts username + otp; also appears to be failing CI/unsupported|
|[passport-totp](https://github.com/jaredhanson/passport-totp)|Presumes a two-phase authentication dance, where the user has already been auth'd via user/pass before this lib validates the totp code|
|[passport-2fa-totp](https://github.com/ilich/passport-2fa-totp)|Also presumes a two-phase dance, and furthermore assumes that the Node app itself is doing the totp validation locally, i.e. has access to the user's second factor secret|

Given that none of these will do the job for us, my next thought was to amend the passport-local strategy to pass through the otp through to Gravity. However there does not appear to be any kind of hook for doing so, thus my proposal to create a forked version that meets our requirements.

### Requirements

- Pass through the triplet of email/password/otp to Gravity to authenticate the user with a single request
- Be indifferent to the absence of the otp, since
  - an app using this strategy may not be enforcing 2fa yet
  - the user logging in may not have enabled 2fa yet
- Return a meaningful error to the call site when 
  - otp is required but missing
  - otp is incorrect

### Is this a good idea?

I'm not positive. Would love feedback. It seems safe to me, and the relevant change is small and has decent test coverage.

**UPDATE 2020-05-05**: see further discussion [here](https://artsy.slack.com/archives/CUG8CB30X/p1588267531426100), [here](https://artsy.slack.com/archives/CUG8CB30X/p1588696300024200) and the `/decision` [here](https://artsy.slack.com/archives/CUG8CB30X/p1588699109034200)

### Whoa, huge diff

Yes, I tried to update documentation, specs and paperwork all in this single PR, leading to a sizable diff.

The heart of the change is actually quite small — [5 LOC basically, excluding specs](https://github.com/artsy/passport-local-with-otp/pull/1/commits/d7bbe41b83e53da433dab1769cf51d6238f50b7b).

### What next?

The remaining followups are tracked in https://artsyproduct.atlassian.net/browse/TRUST-158

### CODE REVIEW

File: lib/strategy.js [hunk 1]
```diff
@@ -45,12 +46,13 @@ function Strategy(options, verify) {
     options = {};
   }
   if (!verify) { throw new TypeError('LocalStrategy requires a verify callback'); }
-  
+
   this._usernameField = options.usernameField || 'username';
   this._passwordField = options.passwordField || 'password';
-  
+  this._totpField = options.totpField || 'totp';
```
Comment (anandaroop) [inline lib/strategy.js]: These strategies allow you to customize the names of the form fields that contain the required fields. I call this simply ~`totp`~ `otp` by default, but our frontend implementation may choose to call it e.g. `otp_attempt` to align with Gravity.

File: lib/strategy.js [hunk 2]
```diff
@@ -45,12 +46,13 @@ function Strategy(options, verify) {
...
+
   passport.Strategy.call(this);
-  this.name = 'local';
+  this.name = 'local-with-totp';
```
Comment (anandaroop) [inline lib/strategy.js]: The strategy needs a unique name so that it can be [invoked](https://github.com/artsy/artsy-passport/compare/master...anandaroop:totp-strategy-support#diff-c7d8f6e9781b93e5f26d9a3fbfbf13c0R20) in the consuming app

File: lib/strategy.js [hunk 3]
```diff
@@ -70,24 +72,25 @@ Strategy.prototype.authenticate = function(req, options) {
   options = options || {};
   var username = lookup(req.body, this._usernameField) || lookup(req.query, this._usernameField);
   var password = lookup(req.body, this._passwordField) || lookup(req.query, this._passwordField);
-  
+  var totp = lookup(req.body, this._totpField) || lookup(req.query, this._totpField);
+
   if (!username || !password) {
     return this.fail({ message: options.badRequestMessage || 'Missing credentials' }, 400);
   }
```
Comment (anandaroop) [inline lib/strategy.js]: This strategy already complains on missing username or password.

But we _don't_ add a similar failure for missing totp here, because it may legitimately not be required depending on the upstream state of `ClientApplication#enforces_otp?` and `User#second_factor_enabled?`

File: lib/strategy.js [hunk 4]
```diff
@@ -70,24 +72,25 @@ Strategy.prototype.authenticate = function(req, options) {
...
-  
+
   try {
     if (self._passReqToCallback) {
-      this._verify(req, username, password, verified);
+      this._verify(req, username, password, totp, verified);
     } else {
-      this._verify(username, password, verified);
+      this._verify(username, password, totp, verified);
```
Comment (anandaroop) [inline lib/strategy.js]: If the ~totp~ otp is present, just pass it through. That's really it!

File: test/strategy.fields.test.js [hunk 5]
```diff
@@ -6,18 +6,18 @@ var chai = require('chai')
 
 
 describe('Strategy', function() {
-    
+
   describe('handling a request with valid credentials in body using custom field names', function() {
```
Comment (anandaroop) [inline test/strategy.fields.test.js:10]: A spec to verify that the ~totp~ otp field can be renamed by the consuming app

File: test/strategy.normal.test.js [hunk 6]
```diff
@@ -6,88 +6,168 @@ var chai = require('chai')
...
-    it('should supply user', function() {
-      expect(user).to.be.an.object;
-      expect(user.id).to.equal('1234');
-    });
-    
-    it('should supply info', function() {
-      expect(info).to.be.an.object;
-      expect(info.scope).to.equal('read');
+
+  describe('when totp needs to be verified', function() {
```
Comment (anandaroop) [inline test/strategy.normal.test.js]: Specs to verify that the strategy works both when the ~totp~ otp is required…

File: test/strategy.normal.test.js [hunk 7]
```diff
@@ -6,88 +6,168 @@ var chai = require('chai')
...
+        expect(info.scope).to.equal('read');
+      });
     });
-    
-    it('should supply user', function() {
-      expect(user).to.be.an.object;
-      expect(user.id).to.equal('1234');
+  })
+
+  describe('when totp does not need to be verified', function() {
```
Comment (anandaroop) [inline test/strategy.normal.test.js]: … and when the ~totp~ otp isn't required.

File: test/strategy.normal.test.js [hunk 8]
```diff
@@ -6,88 +6,168 @@ var chai = require('chai')
...
+        expect(info).to.be.an.object;
+        expect(info.scope).to.equal('read');
+      });
     });
-    
-    it('should supply info', function() {
-      expect(info).to.be.an.object;
-      expect(info.scope).to.equal('read');
+
+    describe('handling a request with valid username and password (but no totp) in query', function() {
```
Comment (anandaroop) [inline test/strategy.normal.test.js]: Spec to verify that an absent ~totp~ otp is not cause for error

**Review by dblandin — APPROVED** [review]
LGTM!

One organizational suggestion: I think it would be nice to organize the commits so that we have:

- 1 commit for initial fork prep: Update README + package.json, remove issue templates, funding file, etc
- 1 commit with core fork changes: minimal src/test updates
- 1 commit with related README/inline documentation updates

What do you think? I think it would nice to be able to link to one commit which includes the primary diff (2nd commit).

Not a blocker though. Feel free to take or leave. :+1:

**Comment by anandaroop** [comment]
I went a little more granular than your suggestion but definitely a good call to disentangle some of those coupled changes and have tidy commits for the crucial implementation change (d7bbe41) and the relevant tests (fd388c3, 47db1f0).

This is identical to the earlier diff apart from whitespace and one stray semicolon. Gonna give this a merge ✌️

**Comment by artsyit** [comment]
:rocket: PR was released in [`v0.3.1`](https://github.com/artsy/passport-local-with-otp/releases/tag/v0.3.1) :rocket:
